python/shotgun_model/data_handler_find.py

```diff
@@ -0,0 +1,384 @@
+                            # shot data dict.
+                            if not self._sg_compare_data(current_record.get(field_name), sg_item.get(field_name)):
+                                diff_list.append({
+                                    "data": ShotgunDataItem(item),
+                                    "mode": self.UPDATED
+                                })
+                                num_modifications += 1
+
+                    # recurse down to the next level
+                    sub_tree = sub_tree[self.CACHE_CHILDREN][unique_field_value]
+
+        # now figure out if anything has been removed
+        self._log_debug("Diffing new tree against old tree...")
+
+        current_uids = set(self._cache[self.CACHE_BY_UID].keys())
+        new_uids = set(new_cache[self.CACHE_BY_UID].keys())
+
+        for deleted_uid in current_uids.difference(new_uids):
+            item = self._cache[self.CACHE_BY_UID][deleted_uid]
+            diff_list.append({
+                "data": ShotgunDataItem(item),
+                "mode": self.DELETED
+            })
+            num_deletes += 1
+
+        # lastly swap the new for8 the old
+        self._clear_cache()
+
+        # at this point, kick the gc to make sure the memory is freed up
+        # despite its cycles.
+        gc.collect()
+
+        # and set the new cache
+        self._cache = new_cache
+
+        self._log_debug("Shotgun data (%d records) received and processed. " % len(sg_data))
+        self._log_debug("    The new tree is %d records." % len(self._cache[self.CACHE_BY_UID]))
+        self._log_debug("    There were %d diffs from in-memory cache:" % len(diff_list))
+        self._log_debug("    Number of new records: %d" % num_adds)
+        self._log_debug("    Number of deleted records: %d" % num_deletes)
+        self._log_debug("    Number of modified records: %d" % num_modifications)
+
+        return diff_list
+
+    def __generate_unique_key(self, parent_unique_key, field, sg_data):
+        """
+        Generates a unique key from a shotgun field.
+
+        Used in conjunction with the hierarchical nature of the find data handler.
+
+        :param parent_unique_key: uid for the parent
+        :param field: the shotgun field that the node represents
+        :param sg_data: associated shotgun data dictionary
+        :returns: Unique string or int
+        """
+        # note: these ids are written to disk and kept in memory
+        # and thus affect memory usage and i/o peformance. We assume
+        # that a vast majority of nodes are leaves and store these as
+        # ints for compactness. non-leaf nodes have a "path-like" string
+        # to uniquely describe their location in the tree.
+        #
+        # we assume that on each level, values are unique.
+
+        value = sg_data.get(field)
+
+        if isinstance(value, dict) and "id" in value and "type" in value:
```

Comment:
i think if you have a polymorphic join, there is a problem here. thanks for the heads up!

so for example if you show tasks grouped by link, you could have tasks for assets and shots in the same list, generating the same ids!